### DISCUSSION TITLE
Files in assets without hash in their names?

### DISCUSSION THREAD

**Discussion by CanadianBeaver** [facebook/docusaurus#10071] [Q&A]
Is it possible to put some files in assets folder without any changes in their names? F.e. I have some .zip files and link from markdown to them like:
`[download](../myfiles/myfile.zip)`
but the link to this file is changing every time when content of file changed like:
`myfile-4fb9ebecd28b6fe6d14e167d763a6f7.zip`
 I prefer to have a permanent link to this file.

**Comment by slorber** [comment]
You can put the file in `static/myfiles/myfile.zip`

Then link to it in 2 ways:

```md
[download](/myfiles/myfile.zip)

[download](pathname:///myfiles/myfile.zip)
```

The 1st option will use our regular asset pipeline, adding a hash so that you can cache requests aggressively (note, your file will still exist at `https://mySite.com/myfiles/myfile.zip`, all static files are served as is, and we only create a new hashed copy)

The 2nd option will opt out of our asset pipeline and leave the link 100% unmodified

**Comment by CanadianBeaver** [comment]
Using `static` folder is not compatible with references in markdown on GitHub

**Reply by slorber** [reply]
> Using static folder is not compatible with references in markdown on GitHub

I'm not sure what you mean here by "references in markdown", please explain with a concrete example

You can also use `<a href="../myfiles/myfile.zip">` and we won't do anything with that link, it will be rendered as is.

**Reply by CanadianBeaver** [reply]
I mean, that if i have in markdown in `doc` folder something like:

`[my file](../files/myfile.zip)`

this file will be accessible on github, when somebody see that markdown. But if I use `static` feature, then link in markdown on github is wrong.

**Reply by slorber** [reply]
If you care about having a stable permalink url with no hash inside, then you could just harcode the link including the domain instead of trying to use a relative path.

```md
[my file](https://myDocusaurusSite.com/files/myfile.zip)
```

If you don't care about having a stable permalink, then the fact there is a hash in the URL shouldn't matter

**Reply by CanadianBeaver** [reply]
This is the worst option. Firstly, there is a strongly dependence on the domain in markdown files. Secondly, the link on github's markdown will lead not to the file in repository, but to a file on the domain.

**Reply by CanadianBeaver** [reply]
the best working option could be something like:

```
  presets: [
    [
      'classic',
      {
        docs: {
          path: '../docs',
          exclude: ['README.md'],
          excludehash: ['*zip'] // HERE
```

**Reply by slorber** [reply]
> This is the worst option. Firstly, there is a strongly dependence on the domain in markdown files. 

Yes, that's the point of a permalink, it includes the domain on purpose.

> Secondly, the link on github's markdown will lead not to the file in repository, but to a file on the domain.

I see no reason why you would want a permalink and yet have your github using a github.com link while your site uses a site.com link. If you want to have permalinks, pick one or another option and hardcode the link.

> the best working option could be something like:

I'm sorry but I'm not going to add a new config option for a niche use case only one user asks for with unconvincing motivations. We will only ever consider this if multiple users complain about it.

If you are looking for a workaround, you could keep this link in your source:

```md
[download](../myfiles/myfile.zip) 
```

And then write a small remark plugin so that inside your Docusaurus site, it gets "rewritten" to 

```md
[download](../myfiles/myfile.zip)
```

This is pseudo code but something like this should work:

```ts
export default function plugin(options) {
  return async (root) => {
    const {visit} = await import('unist-util-visit');
    visit(root, 'link', (link: Link) => {
      if (link.url === "../myfiles/myfile.zip") {
        link.url = "pathname:///myfiles/myfile.zip"
      }
    });
  };
}
```

**Reply by CanadianBeaver** [reply]
Thank you, I tried the suggested option and it may help, but it seems like a workaround with static files and self-made plugin. Is there really no other option to have permanent links to files located in the `assets/files` folder (without a hash)?

I don't think this is unique to my project. I believe that projects that distribute files would also like to be able to follow links in .md files (github, gitlab boards) and permalinks on the generated website.

**Reply by slorber** [reply]
No we don't have any other option. Most users are fine to have a hash and it improves caching.

There can only be one Markdown link syntax that works with GitHub. And when it's used on Docusaurus, we either add the hash, or we don't. 

We choose to add a hash (which most users will be fine with), and provide alternatives/workarounds for users to opt-out.

If we choose to **not** add an hash, then we'll also have to provide similar alternatives/workarounds for users that do want the hash. It would be the same problem, from a different perspective.

---

> I don't think this is unique to my project. I believe that projects that distribute files would also like to be able to follow links in .md files (github, gitlab boards) and permalinks on the generated website.

Afaik this has never been requested before.

If you have a permalink, the domain is part of that permalink. 
Here you want to have one permalink for GitHub, and one for Docusaurus. 
What's the point of having 2 distinct URLs targeting the same file? 
What's the point of having GitHub users downloading from GitHub, and website visitors downloading from the website? 

If you want a permalink, then don't be afraid to couple to a domain, because the domain is part of it. 
It will also make things more explicit and ensure you don't inadvertently break the permalink by refactoring your docs.

**Comment by doormat18** [comment]
Just wanted to add my voice to this, been trying to figure out how to do this for the last hour. So there are certainly multiple people trying to do this.

**Reply by charlespalmerbf** [reply]
Just in case this helps you, the following resolved the issue I was having where I have a cross-domain deployed site that needed to serve without the appended hash.

Adding the pathname:// seemed to sort it for my usecase.

`[link](pathname:///img/filename_here.svg)`

**Reply by doormat18** [reply]
Did you have the serve the file out of the static directory?

**Reply by charlespalmerbf** [reply]
Sadly yes, ideally that wouldn't have to be the case but there didn't seem to be another way to do it, and at the end of the day, for my usecase it did end up working okay.

**Reply by doormat18** [reply]
That worked! Thanks @charlespalmerbf

**Reply by charlespalmerbf** [reply]
No worries, glad it sorted it out.
